### PR TITLE
Adds validation to check if there is any invalid offer when checkout …

### PR DESCRIPTION
**Pull request by patrickemuller**
…with transfer

@ssjr cara, da uma olhada, eu queria ter feito usando Pundit mas não sei como faz pra passar objetos customizados pra essa por**.

### CODE REVIEW

File: spec/support/blueprints.rb [hunk 1]
```diff
@@ -96,9 +113,23 @@
   total { 1_000_00 }
 end
 
+Purchase.blueprint(:invalid) do
```
Comment (ssjr) [inline spec/support/blueprints.rb]: pending = invalid? what's the difference of the blueprint above?

File: app/controllers/checkouts_controller.rb [hunk 2]
```diff
@@ -5,6 +5,12 @@ class CheckoutsController < ApplicationController
   # Used to show the page to upload the transfer receipt
   def transfer
     @purchase = Purchase.find_by(invoice_id: params[:invoice_id])
+
+    if @purchase.has_invalid_offers?
```
Comment (ssjr) [inline app/controllers/checkouts_controller.rb]:
``` ruby
# app/controllers/checkouts_controller.rb
before_action :authenticate_user!, only: %i(checkout process_payment transfer)
rescue_from Pundit::NotAuthorizedError, with: :cant_pay_anymore, only: :transfer

def transfer
  @purchase = Purchase.find_by(invoice_id: params[:invoice_id])
  authorize @purchase
  @bank_account = BankAccount.first
end

private

def cant_pay_anymore
  redirect_to root_path, flash: { alert: 'Acesso negado. Talvez uma oferta desta compram já foi encerrada' }
end
```

``` ruby
# app/policies/purchase_policy.rb

def transfer?
  record.has_invalid_offers? && user == record.user
end
```

File: spec/support/blueprints.rb (hunk 1 above)
Comment (patrickemuller) [inline spec/support/blueprints.rb]: In fact what make this Purchase invalid is the fact that some offer related to this is expired, nothing more.

Comment (patrickemuller) [inline spec/support/blueprints.rb]: In fact what make this Purchase invalid is the fact that some offer related to this is expired, nothing more.

**Pull request merged**